samples/host_exerciser/host_exerciser.h

```diff
@@ -0,0 +1,489 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+#include <opae/cxx/core/events.h>
+#include <opae/cxx/core/shared_buffer.h>
+
+#include "afu_test.h"
+
+const char *AFU_ID = "91c2a3a1-4a23-4e21-a7cd-2b36dbf2ed73";
+
+namespace host_exerciser {
+using opae::fpga::types::event;
+using opae::fpga::types::shared_buffer;
+
+static const uint64_t HELPBK_TEST_TIMEOUT = 30000;
+static const uint64_t CL = 64;
+static const uint64_t KB = 1024;
+static const uint64_t MB = KB * 1024;
+static const uint64_t LOG2_CL = 6;
+static const size_t LPBK1_DSM_SIZE = 2 * MB;
+static const size_t LPBK1_BUFFER_SIZE = 1 * MB;
+static const size_t LPBK1_BUFFER_ALLOCATION_SIZE = 2 * MB;
+static const uint32_t DSM_STATUS_TEST_COMPLETE = 0x40;
+
+// Host execiser CSR Offset
+enum {
+  DFH_HEADER	= 0x0000,
+  DFH_CAPABILITY_HEADER = 0x0008,
+  DFH_CAPABILITY_GUID = 0x0010,
+  DFH_CAPABILITY_INT = 0x0018,
+  DSM_STATUS = 0x0040,
+  CSR_SCRATCHPAD0 = 0x0100,
+  CSR_SCRATCHPAD1 = 0x0108,
+  CSR_AFU_DSM_BASEL = 0x0110,
+  CSR_AFU_DSM_BASEH = 0x0114,
+  CSR_SRC_ADDR = 0x0120,
+  CSR_DST_ADDR = 0x0128,
+  CSR_NUM_LINES = 0x0130,
+  CSR_CTL = 0x0138,
+  CSR_CFG = 0x0138,
+  CSR_INACT_THRESH = 0x0148,
+  CSR_INTERRUPT0 = 0x0150,
+  CSR_SWTEST_MSG = 0x0158,
+  CSR_STATUS0 = 0x0160,
+  CSR_STATUS1 = 0x0168,
+  CSR_ERROR = 0x0170,
+  CSR_ERROR1 = 0x0178
+};
+
+//configures test mode
+typedef enum {
+	HOST_EXEMODE_LPBK1= 0x0,
+	HOST_EXEMODE_READ = 0x1,
+	HOST_EXEMODE_WRITE= 0x2,
+	HOST_EXEMODE_TRUPT = 0x3,
+} host_exe_mode;
+
+//read request type
+typedef enum {
+	HOSTEXE_RDLINE_S = 0x0,
+	HOSTEXE_RDLINE_L = 0x1,
+	HOSTEXE_RDLINE_MIX = 0x2,
+} hostexe_rd;
+
+//read request type
+typedef enum {
+	HOSTEXE_WRLINE_M = 0x0,
+	HOSTEXE_WRLINE_L = 0x1,
+} hostexe_wr;
+
+// DFH Header
+union dfh_header  {
+  enum {
+    offset = DFH_HEADER
+  };
+
+  dfh_header(uint64_t v) : value(v) {}
+  uint64_t value;
+  struct {
+    uint16_t FeatureID : 12;
+    uint8_t FeatureRev : 4;
+    uint32_t NextDfhByteOffset : 24;
+	uint8_t EOL : 1;
+	uint8_t Reserved41 : 7;
+	uint8_t AFUminor : 4;
+	uint8_t DFHversion : 8;
+	uint8_t FeatureType : 4;
+  };
+};
+
+// DFH CAPABILITY Header
+union dfh_cap_hdr {
+	enum {
+		offset = DFH_CAPABILITY_HEADER
+	};
+	dfh_cap_hdr(uint32_t v) : value(v) {}
+	uint32_t value;
+	struct {
+		uint16_t CapID : 16;
+		uint8_t CapIDVersion : 4;
+		uint16_t NextCapOffset : 12;
+	};
+};
+
+// TBD
+union dfh_cap_guid {
+	enum {
+		offset = DFH_CAPABILITY_GUID
+	};
+
+	dfh_cap_guid(uint8_t* v)
+	{
+		for (int i = 0; i < 24; i++) {
+			memcpy(&value[i], v, 1);
+			v++;
+		}
+	}
+	uint8_t  value[24];
+	struct {
+		uint64_t  GuidHi;
+		uint64_t GuidLo;
+		uint32_t Reserved;
+		uint32_t dfhCapHeader;
+	};
+};
+
+
+union dfh_cap_init {
+	enum {
+		offset = DFH_CAPABILITY_INT
+	};
+
+	dfh_cap_init(uint8_t* v)
+	{
+		for (int i = 0; i < 20; i++) {
```

Comment:
sizeof(value) to be safe?

Reply:
defined structure  dfh_cap_init_s and remove memory copy. fixed